### PR TITLE
[CALCITE-1338] JoinProjectTransposeRule makes wrong transformation when the right child of left outer join has a RexLiteral project expression

### PR DESCRIPTION
**Pull request by chunweilei**
JIRA: https://issues.apache.org/jira/browse/CALCITE-1338.

### CODE REVIEW

**Comment by zhztheplayer** [comment]
Since the JIRA is opened for a bug, could you please change the commit message to describe the bug directly? Thanks.

**Comment by chunweilei** [comment]
@zhztheplayer Done. Thanks for your reminder.

**Comment by julianhyde** [comment]
In future, when you make code changes after a review, could you add new commits rather than amending existing commits? It's easier to review the deltas.

(Obviously you need to force-push in order to rebase or change the commit message.)

As a reviewer I am happy to squash and rebase (as long as rebase is clean or relatively easy).

I'm re-reviewing now, hopefully will be able to commit to master, so don't touch the branch.

**Comment by chunweilei** [comment]
Sure. Thank you for your kindly reminder @julianhyde .

File: core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java [hunk 1]
```diff
@@ -510,6 +510,118 @@ private void basePushFilterPastAggWithGroupingSets(boolean unchanged)
...
+    final HepProgram program =
+        HepProgram.builder()
+            .addRuleInstance(JoinProjectTransposeRule.RIGHT_PROJECT_INCLUDE_OUTER)
+            .build();
+    final String sql = "select *\n"
+        + "from dept a\n"
+        + "left join (select x name, y is not null from\n"
+        + "(values (2, cast(null as integer)), (2, 1)) as t(x, y)) b\n"
+        + "on a.name = b.name";
+    checkPlanUnchanged(new HepPlanner(program), sql);
```
Comment (jinxing64) [inline core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java]: Hi @chunweilei , I'm learning this PR ~
In the right project of this left-join, the field of `name` is of Integer type. Do you really mean this?

**Pull request closed without merging**